File: backend/internal/handlers/book.go
package handlers

import (
	"github.com/gin-gonic/gin"
	"net/http"
	"online_library/backend/internal/middleware"
	"online_library/backend/internal/models"
	"online_library/backend/internal/pkg/roles"
	"online_library/backend/internal/service"
	"strconv"
)

type BookHandler struct {
	bookService service.BookService
}

type TagListRequest struct {
	TagIDs []int `json:"tag_ids"`
}

type AuthorListRequest struct {
	AuthorIDs []int `json:"author_ids"`
}

type StatusUpdateRequest struct {
	Status string `json:"status"`
}

func NewBookHandler(bookService service.BookService) *BookHandler {
	return &BookHandler{bookService: bookService}
}

func (h *BookHandler) CreateBook(c *gin.Context) {
	userID, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	var book models.Book
	if err := c.ShouldBindJSON(&book); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid JSON"})
		return
	}

	bookID, err := h.bookService.CreateBook(&book, userRole, userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusCreated, gin.H{"id": bookID})
}

func (h *BookHandler) UpdateBook(c *gin.Context) {
	userID, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	var book models.Book
	if err := c.ShouldBindJSON(&book); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid JSON"})
		return
	}

	if err := h.bookService.UpdateBook(&book, userID, userRole); err != nil {
		c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusOK)
}

func (h *BookHandler) DeleteBook(c *gin.Context) {
	userID, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	bookID, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid book ID"})
		return
	}

	if err := h.bookService.DeleteBook(bookID, userID, userRole); err != nil {
		c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusOK)
}

func (h *BookHandler) GetBookByID(c *gin.Context) {
	userID, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	bookID, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid book ID"})
		return
	}

	book, err := h.bookService.GetBookByID(bookID, userID, userRole)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, book)
}

func (h *BookHandler) GetBooksByStatuses(c *gin.Context) {
	_, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	offset, _ := strconv.Atoi(c.DefaultQuery("offset", "0"))
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "20"))

	books, err := h.bookService.GetBooksByStatuses(userRole, offset, limit)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, books)
}

func (h *BookHandler) SearchBooks(c *gin.Context) {
	_, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	query := c.Query("q")
	offset, _ := strconv.Atoi(c.DefaultQuery("offset", "0"))
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "20"))

	books, err := h.bookService.SearchBooks(query, userRole, limit, offset)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, books)
}

func (h *BookHandler) GetBooksByAuthor(c *gin.Context) {
	_, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	authorID, err := strconv.Atoi(c.Param("author_id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid author ID"})
		return
	}

	offset, _ := strconv.Atoi(c.DefaultQuery("offset", "0"))
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "20"))

	books, err := h.bookService.GetBooksByAuthor(authorID, userRole, offset, limit)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, books)
}

func (h *BookHandler) GetBooksByTag(c *gin.Context) {
	_, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	tagID, err := strconv.Atoi(c.Param("tag_id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid tag ID"})
		return
	}

	offset, _ := strconv.Atoi(c.DefaultQuery("offset", "0"))
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "20"))

	books, err := h.bookService.GetBooksByTag(tagID, userRole, offset, limit)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, books)
}

func (h *BookHandler) GetUserBooks(c *gin.Context) {
	userID, _, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	books, err := h.bookService.GetUserBooks(userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, books)
}

func (h *BookHandler) GetUserFavoriteBooks(c *gin.Context) {
	userID, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	books, err := h.bookService.GetUserFavoriteBooks(userID, userRole)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, books)
}

func (h *BookHandler) AddBookToFavorites(c *gin.Context) {
	userID, _, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	bookID, err := strconv.Atoi(c.Param("book_id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid book ID"})
		return
	}

	if err := h.bookService.AddBookToFavorites(userID, bookID); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusNoContent)
}

func (h *BookHandler) RemoveBookFromFavorites(c *gin.Context) {
	userID, _, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	bookID, err := strconv.Atoi(c.Param("book_id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid book ID"})
		return
	}

	if err := h.bookService.RemoveBookFromFavorites(userID, bookID); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusNoContent)
}

func (h *BookHandler) SetBookTags(c *gin.Context) {
	userID, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	bookID, err := strconv.Atoi(c.Param("book_id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid book ID"})
		return
	}

	var req TagListRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid JSON"})
		return
	}

	if err := h.bookService.SetBookTags(bookID, req.TagIDs, userID, userRole); err != nil {
		c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusNoContent)
}

func (h *BookHandler) AddBookTag(c *gin.Context) {
	userID, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	bookID, _ := strconv.Atoi(c.Param("book_id"))
	tagID, _ := strconv.Atoi(c.Param("tag_id"))

	if err := h.bookService.AddBookTag(bookID, tagID, userID, userRole); err != nil {
		c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusNoContent)
}

func (h *BookHandler) RemoveBookTag(c *gin.Context) {
	userID, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	bookID, _ := strconv.Atoi(c.Param("book_id"))
	tagID, _ := strconv.Atoi(c.Param("tag_id"))

	if err := h.bookService.RemoveBookTag(bookID, tagID, userID, userRole); err != nil {
		c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusNoContent)
}

func (h *BookHandler) SetBookAuthors(c *gin.Context) {
	userID, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	bookID, _ := strconv.Atoi(c.Param("book_id"))

	var req AuthorListRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid JSON"})
		return
	}

	if err := h.bookService.SetBookAuthors(bookID, req.AuthorIDs, userID, userRole); err != nil {
		c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusNoContent)
}

func (h *BookHandler) AddBookAuthor(c *gin.Context) {
	userID, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	bookID, _ := strconv.Atoi(c.Param("book_id"))
	authorID, _ := strconv.Atoi(c.Param("author_id"))

	if err := h.bookService.AddBookAuthor(bookID, authorID, userID, userRole); err != nil {
		c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusNoContent)
}

func (h *BookHandler) RemoveBookAuthor(c *gin.Context) {
	userID, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	bookID, _ := strconv.Atoi(c.Param("book_id"))
	authorID, _ := strconv.Atoi(c.Param("author_id"))

	if err := h.bookService.RemoveBookAuthor(bookID, authorID, userID, userRole); err != nil {
		c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusNoContent)
}

func (h *BookHandler) UpdateBookStatus(c *gin.Context) {
	_, userRole, ok := middleware.ExtractUser(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	bookID, _ := strconv.Atoi(c.Param("book_id"))

	var req StatusUpdateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid JSON"})
		return
	}

	if err := h.bookService.UpdateBookStatus(bookID, req.Status, userRole); err != nil {
		c.JSON(http.StatusForbidden, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusNoContent)
}

func (h *BookHandler) GetDuplicateBooks(c *gin.Context) {
	_, role, ok := middleware.ExtractUser(c)
	if !ok || !roles.IsAdmin(role) {
		c.JSON(http.StatusForbidden, gin.H{"error": "admin access required"})
		return
	}

	title := c.Param("title")
	if title == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "missing title"})
		return
	}

	books, err := h.bookService.GetDuplicateBooks(title)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to find duplicates"})
		return
	}

	c.JSON(http.StatusOK, books)
}

File: frontend/src/api/book.js
import axios from '@/api/axios'

/**
 * Получает информацию о книге по ID.
 * @param {number|string} id - ID книги
 * @returns {Promise<Object>} Данные книги
 */
export async function getBook(id) {
    const { data } = await axios.get(`/books/${id}`)
    return data
}

/**
 * Получает дополнительные данные о книге (например, оценки, количество комментариев и т.п.)
 * @param {number|string} id - ID книги
 * @returns {Promise<Object>} Дополнительная информация
 */
export async function getBookExtras(id) {
    const { data } = await axios.get(`/books/${id}/extras`)
    return data
}

/**
 * Ищет книги с параметрами.
 * @param {Object} params - query параметры поиска (title, author и т.п.)
 * @returns {Promise<Array>} Результаты поиска
 */
export async function searchBooks(params = {}) {
    const { data } = await axios.get('/books', { params })
    return data
}

/**
 * Получает книги по автору.
 * @param {number|string} authorId
 * @returns {Promise<Array>}
 */
export async function getBooksByAuthor(authorId) {
    const { data } = await axios.get(`/books/author/${authorId}`)
    return data
}

/**
 * Получает книги по тегу.
 * @param {number|string} tagId
 * @returns {Promise<Array>}
 */
export async function getBooksByTag(tagId) {
    const { data } = await axios.get(`/books/tag/${tagId}`)
    return data
}

/**
 * Проверяет наличие дубликатов книги по названию.
 * @param {string} title
 * @returns {Promise<Array>}
 */
export async function getDuplicateBooks(title) {
    const { data } = await axios.get(`/books/duplicates/${encodeURIComponent(title)}`)
    return data
}

/**
 * Получает книги текущего пользователя.
 * @returns {Promise<Array>}
 */
export async function getUserBooks() {
    const { data } = await axios.get('/books/mine')
    return data
}

/**
 * Получает новые релизы книг.
 * @returns {Promise<Array>}
 */
export async function getNewReleases() {
    const { data } = await axios.get('/books/new-releases')
    return data
}

/**
 * Получает избранные книги пользователя.
 * @returns {Promise<Array>}
 */
export async function getFavoriteBooks() {
    const { data } = await axios.get('/books/favorites')
    return data
}

/**
 * Добавляет книгу в избранное.
 * @param {number|string} bookId
 * @returns {Promise<Object>}
 */
export async function addToFavorites(bookId) {
    const { data } = await axios.post(`/books/${bookId}/favorite/add`)
    return data
}

/**
 * Удаляет книгу из избранного.
 * @param {number|string} bookId
 * @returns {Promise<Object>}
 */
export async function removeFromFavorites(bookId) {
    const { data } = await axios.post(`/books/${bookId}/favorite/remove`)
    return data
}

/**
 * Создаёт новую книгу.
 * @param {Object} bookData
 * @returns {Promise<Object>}
 */
export async function createBook(bookData) {
    const { data } = await axios.post('/books', bookData)
    return data
}

/**
 * Обновляет книгу (владелец или админ).
 * @param {number|string} bookId
 * @param {Object} updates
 * @returns {Promise<Object>}
 */
export async function updateBook(bookId, updates) {
    const { data } = await axios.post(`/books/${bookId}`, updates)
    return data
}

/**
 * Удаляет книгу (владелец или админ).
 * @param {number|string} bookId
 * @returns {Promise<Object>}
 */
export async function deleteBook(bookId) {
    const { data } = await axios.post(`/books/${bookId}/delete`)
    return data
}

/**
 * Обновляет статус книги (только админ).
 * @param {number|string} bookId
 * @param {Object} statusData
 * @returns {Promise<Object>}
 */
export async function updateBookStatus(bookId, statusData) {
    const { data } = await axios.post(`/books/${bookId}/status`, statusData)
    return data
}

/**
 * Устанавливает авторов книги (заменяет всех).
 * @param {number|string} bookId
 * @param {Array<number>} authorIds
 * @returns {Promise<Object>}
 */
export async function setBookAuthors(bookId, authorIds) {
    const { data } = await axios.post(`/books/${bookId}/authors`, { author_ids: authorIds })
    return data
}

/**
 * Добавляет автора к книге.
 * @param {number|string} bookId
 * @param {number|string} authorId
 * @returns {Promise<Object>}
 */
export async function addBookAuthor(bookId, authorId) {
    const { data } = await axios.post(`/books/${bookId}/authors/${authorId}`)
    return data
}

/**
 * Удаляет автора из книги.
 * @param {number|string} bookId
 * @param {number|string} authorId
 * @returns {Promise<Object>}
 */
export async function removeBookAuthor(bookId, authorId) {
    const { data } = await axios.post(`/books/${bookId}/authors/${authorId}/remove`)
    return data
}

/**
 * Устанавливает теги книги (заменяет все).
 * @param {number|string} bookId
 * @param {Array<number>} tagIds
 * @returns {Promise<Object>}
 */
export async function setBookTags(bookId, tagIds) {
    const { data } = await axios.post(`/books/${bookId}/tags`, { tag_ids: tagIds })
    return data
}

/**
 * Добавляет тег к книге.
 * @param {number|string} bookId
 * @param {number|string} tagId
 * @returns {Promise<Object>}
 */
export async function addBookTag(bookId, tagId) {
    const { data } = await axios.post(`/books/${bookId}/tags/${tagId}`)
    return data
}

/**
 * Удаляет тег из книги.
 * @param {number|string} bookId
 * @param {number|string} tagId
 * @returns {Promise<Object>}
 */
export async function removeBookTag(bookId, tagId) {
    const { data } = await axios.post(`/books/${bookId}/tags/${tagId}/remove`)
    return data
}